### DISCUSSION TITLE
Getting a formula (isl) to use current version of GCC, and building gcc on High Sierra from now on

### DISCUSSION THREAD

**Discussion by Jerry1144** [Homebrew/discussions#3513] [Tap maintenance and Homebrew development] [answered]
With `isl` hitting version 0.25, it now uses C++17 features and fails to build on High Sierra, whose CLT10.14 clang correctly reports that this is not supported on this 10.13 system:
```
./configure --prefix=/usr/local/Cellar/isl/0.25 --with-gmp=system --with-gmp-prefix=/usr/local/opt/gmp
make
(snip)
clang++ -std=c++11 -std=gnu++17 -DHAVE_CONFIG_H   -I. -I. -I./include -Iinclude/ -I/usr/local/opt/gmp/include   -g -O2 -c -o isl_test_cpp17.o isl_test_cpp17.cc
In file included from isl_test_cpp17-checked.cc:23:
./isl_test_cpp17-generic.cc:48:16: error: call to unavailable member function 'value': introduced in macOS 10.14
        if (maybe_int.value() != 5)
            ~~~~~~~~~~^~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:942:27: note: candidate function has been explicitly made unavailable
    constexpr value_type& value() &
                          ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:933:33: note: candidate function has been explicitly made unavailable
    constexpr value_type const& value() const&
                                ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:951:28: note: candidate function not viable: no known conversion from 'optional<...>' to 'optional<...>' for object argument
    constexpr value_type&& value() &&
                           ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:960:34: note: candidate function not viable: no known conversion from 'optional<...>' to 'const optional<...>' for object argument
    constexpr value_type const&& value() const&&
                                 ^
(same error also occurs in isl_test_cpp17-checked.cc, which gets compiled next)
```
Naturally I tried to assign a C++17-compatible compiler at this formula. However, this update coincided with `gcc` upgrading to 12.1.0, which unfortunately depends on `isl` by itself… So I need to somehow make do with the `gcc 11.3.0_1` I currently have installed. I'm asking for help for modifying `Formula/isl.rb`, so that it compiles with minimal interference for all future upgrades. I want to add a build dependency block for High Sierra and below.

```Section break```

I now report what I have tried.

My version is migrated to `gcc@11` at version 11.3.0, so I `brew unlink gcc`, ~renamed `Cellar/gcc/` to `Cellar/gcc@11/` and then `Cellar/gcc@11/11.3.0_1/` to `Cellar/gcc@11/11.3.0/`, and finally `brew link gcc@11`~ made a symlink from gcc@11/11.3.0 to gcc/11.3.0_1 because moving the executables broke them.

- **Passing --cc=gcc-11 directly (viz. `brew install isl --cc=gcc-11`): Success.** But then I'll have to keep this old version of gcc around, or symlink the newest gcc every time, because `brew upgrade` keeps previous install options.
- Passing --cc=gcc: Failure. Enigmatically, this doesn't require me to upgrade my `gcc` first. It still fails for compiler not supporting C++17's optionals (the same errors I pasted above), so it's probably not using Homebrew's gcc.
- Adding `depends_on "gcc@11" => :build` to `isl.rb`: Failure in all circumstances below for the same error. Apparently the gcc it found was still the alias to system `clang`
  - Upgrading directly
  - Setting `CC=gcc` in the interactive build environment
  - Passing --cc=gcc
- Adding `depends_on "gcc" => :build` instead: Failure. This time Homebrew wants to upgrade gcc first, which causes a circular dependency: ```==> Installing dependencies for isl: gcc
==> Installing isl dependency: gcc
==> Installing dependencies for gcc: isl
==> Installing gcc dependency: isl
==> Installing dependencies for isl: gcc
==> Installing isl dependency: gcc
(snip)
^C```
  - Pulling out the old `gcc.rb` file matching my installed version, so that brew doesn't hit this circular dependency bug: Failure with the same C++17 unsupported error.

All in all, the simplest way to solve this - adding a dependency block - doesn't seem to work, nor does the un-versioned --cc=gcc option. I don't like the working way left, so I'm asking for help.

```Section break```

For all the unfortunate souls stuck on High Sierra, who don't have any version of Homebrew's `gcc` yet, so can't even get the first gcc installed, here's a solution - pull out an old version of `isl`'s formula and install it first (I suppose you have dev tools already):
```
brew update         # Just to minimize the chance the formula gets updated in the mean time
cd "$(brew --repository homebrew/core)"
git checkout 912df90f851785d85995ea89db3ce125d7f4114e -- Formula/isl.rb         # This reverts to isl 0.24 which can be compiled with system clang
brew install isl
brew install gcc@11         # It's a long long build
git checkout master -- Formula/isl.rb         # Get the newest formula back
brew install isl --cc=gcc-11         # I hope some day we can pass --cc=gcc here, but not now
brew install gcc        # now you get the latest gcc
```

**Comment by carlocab** [comment, marked as answer]
One other possibility is to `brew install llvm` and then pass `--cc=llvm_clang` when building ISL and/or GCC.

**Reply by Jerry1144** [reply]
Whoops, I totally forgot about Homebrew's llvm.

**Reply by bettabacon** [reply]
I had great success using this to solve `isl`, but it failed when trying to build `gcc`. Regardless, after building `isl`, I then was able to build `gcc` successfully without `--cc=llvm_clang` (just normal `brew install gcc`). I'm also on macOS 10.13.6 High Sierra.

**Reply by patrikaxelsson** [reply]
On High Sierra here too. I had isl 0.24 installed.

I could not install llvm as it requires z3, which requires isl 0.26. I could however install llvm@15, but using `brew install --cc=llvm_clang isl` then to install isl 0.26 failed: it claimed the compiler could not create executables.

Found another solution inspired by @fengerzh's reply below: gcc@11 also depends on isl 0.26, but in reality gcc@11 does not need it to be built/installed nor to be used, so I was able to install isl 0.26 by:
1. Make sure gcc@11 has all other dependencies installed except isl 0.26: `brew install gmp libmpc mpfr zstd`
2. `brew install --ignore-dependencies gcc@11`
3. `brew install isl --cc=gcc-11`

Now with isl 0.26 built and installed, it is possible to for example do a `brew install gcc`.

**Reply by johhnry** [reply]
On High Sierra / macOS 10.13 too, I can't install LLVM because it tries to install z3: `z3 cannot be built with any available compilers, Install GNU's GCC: brew install gcc`

But then I can't install gcc because I can't install isl ^^

What can I do?

**Reply by fengerzh** [reply]
You can try this: https://github.com/orgs/Homebrew/discussions/4949#discussioncomment-7799075 I followed by installing llvm 12 first, then modify llvm 17 source code, and install successfully.

**Reply by nhwebforge** [reply]
Argh I feel like everthing I try seems like it's going to work and then another error. @patrikaxelsson your idea is great but when I try to install GCC 11 ignoring the dependencies, it trys to patch it using "`Applying gcc-11.4.0.diff`", which in turn returns the error:
```
checking for isl 0.15 or later... no
required isl version is 0.15 or later
configure: error: Unable to find a usable isl.  See config.log for details.
```
Any ideas how to force an install of 11.0 and not patch to 11.4 - basically ignore the .diff file?

Thanks in advance

**Reply by patrikaxelsson** [reply]
@nhwebforge: Your attempt shows that the reason I could install gcc@11 without dependencies was because I already had the older isl 0.24 installed. So if there is a way to install the older isl 0.24, it should enable you to install gcc@11.

**Reply by konnano** [reply]
Save it as you will rewrite (delete) the formula.

cp $(brew --repo)/Library/Taps/homebrew/homebrew-core/Formula/g/gcc.rb ~/gcc.rb

brew edit gcc # disable isl

\# depends_on "isl" # Comment line 52

--with-isl=#{Formula["isl"].opt_prefix} # Delete line 98

brew install gcc

brew install --cc=gcc-13 isl

**Comment by Bo98** [comment]
Hmm, these parts should be optional. C++17 is only used for tests. We should probably look into ways around that.

**Reply by Bo98** [reply]
Passing `ax_cv_cxx_compile_cxx17=no` to `configure` could work as a hack.

**Comment by Jerry1144** [comment]
Tried a `fails_with :clang` block. This time brew correctly called the only `gcc` available without me even specifying a `depends_on`.

Will open a PR.

**Reply by Jerry1144** [reply]
Or not, if we can get around this.

**Comment by markusbkk** [comment]
How do I need to edit the formula (i e `brew edit isl`) to make the build use the homebrew installed LLVM/Clang?

I followed the directions on [this Stackoverflow thread](https://stackoverflow.com/questions/69906053/how-to-install-llvm13-with-homerew-on-macos-high-sierra-10-13-6-got-built-tar/72334166#72334166) and have a functioning LLVM 13 installation with C++17 support.

**Reply by konnano** [reply]
llvm@15 is required for isl builds on 10.13
The installation method of llvm@15 is the same

/usr/local/Homebrew/Library/Homebrew/shims/super/cc on line 80

"#{ENV["HOMEBREW_PREFIX"]}/opt/llvm/bin/#{Regexp.last_match(1)}"

rewrite this as

"#{ENV["HOMEBREW_PREFIX"]}/opt/llvm@15/bin/#{Regexp.last_match(1)}"

brew install --cc=llvm_clang isl

**Reply by carlocab** [reply]
Try this:
```diff
diff --git a/Formula/isl.rb b/Formula/isl.rb
index 474fb6ba3d3..1a5cf51c437 100644
--- a/Formula/isl.rb
+++ b/Formula/isl.rb
@@ -37,6 +37,7 @@ class Isl < Formula
   depends_on "gmp"
 
   def install
+    ENV["ax_cv_cxx_compile_cxx17"] = "no"
     system "./autogen.sh" if build.head?
     system "./configure", "--disable-dependency-tracking",
                           "--disable-silent-rules",
```
You might be able to build `isl` without needing to install `llvm@*`.

**Reply by markusbkk** [reply]
> llvm@15 is required for isl builds on 10.13 The installation method of llvm@15 is the same
> 
> /usr/local/Homebrew/Library/Homebrew/shims/super/cc on line 80
> 
> "#{ENV["HOMEBREW_PREFIX"]}/opt/llvm/bin/#{Regexp.last_match(1)}"
> 
> rewrite this as
> 
> "#{ENV["HOMEBREW_PREFIX"]}/opt/llvm@15/bin/#{Regexp.last_match(1)}"
> 
> brew install --cc=llvm_clang isl

Thanks! I followed these instructions and upgraded to LLVM 15 (exported the path in `~/.profile`, too).

Unfortunately, even after rewriting the shim to `"#{ENV["HOMEBREW_PREFIX"]}/opt/llvm@15/bin/#{Regexp.last_match(1)}"` I receive `error: C compiler cannot create executables`

**Reply by markusbkk** [reply]
> Try this:
> 
> ```diff
> diff --git a/Formula/isl.rb b/Formula/isl.rb
> index 474fb6ba3d3..1a5cf51c437 100644
> --- a/Formula/isl.rb
> +++ b/Formula/isl.rb
> @@ -37,6 +37,7 @@ class Isl < Formula
>    depends_on "gmp"
>  
>    def install
> +    ENV["ax_cv_cxx_compile_cxx17"] = "no"
>      system "./autogen.sh" if build.head?
>      system "./configure", "--disable-dependency-tracking",
>                            "--disable-silent-rules",
> ```
> 
> You might be able to build `isl` without needing to install `llvm@*`.

Tried that, as well. Doesn't work. I think there's a lot more than just C++17 support that keeps **isl** from compiling on old versions of Clang.

I really wonder why the installation with `cc=llvm_clang` reports a non-functioning C compiler though. I'm already compiling my own software with it and it definitely works.

**Reply by Jerry1144** [reply]
> I really wonder why the installation with `cc=llvm_clang` reports a non-functioning C compiler though. I'm already compiling my own software with it and it definitely works.

Homebrew does not select llvm_clang in any means (we have --cc=gcc-13 but only an unversioned --cc=llvm_clang), instead takes whatever `/usr/local/opt/llvm` points to. I would `ln -s ./llvm@15 /usr/local/opt/llvm` to select llvm@15, or `ln -s /usr/local/Cellar/llvm/17.0.5 /usr/local/opt/llvm` to select the current `llvm` at version 17.0.5.

**Comment by fengerzh** [comment]
> With `isl` hitting version 0.25, it now uses C++17 features and fails to build on High Sierra, whose CLT10.14 clang correctly reports that this is not supported on this 10.13 system:
> 
> ```
> ./configure --prefix=/usr/local/Cellar/isl/0.25 --with-gmp=system --with-gmp-prefix=/usr/local/opt/gmp
> make
> (snip)
> clang++ -std=c++11 -std=gnu++17 -DHAVE_CONFIG_H   -I. -I. -I./include -Iinclude/ -I/usr/local/opt/gmp/include   -g -O2 -c -o isl_test_cpp17.o isl_test_cpp17.cc
> In file included from isl_test_cpp17-checked.cc:23:
> ./isl_test_cpp17-generic.cc:48:16: error: call to unavailable member function 'value': introduced in macOS 10.14
>         if (maybe_int.value() != 5)
>             ~~~~~~~~~~^~~~~
> /Library/Developer/CommandLineTools/usr/include/c++/v1/optional:942:27: note: candidate function has been explicitly made unavailable
>     constexpr value_type& value() &
>                           ^
> /Library/Developer/CommandLineTools/usr/include/c++/v1/optional:933:33: note: candidate function has been explicitly made unavailable
>     constexpr value_type const& value() const&
>                                 ^
> /Library/Developer/CommandLineTools/usr/include/c++/v1/optional:951:28: note: candidate function not viable: no known conversion from 'optional<...>' to 'optional<...>' for object argument
>     constexpr value_type&& value() &&
>                            ^
> /Library/Developer/CommandLineTools/usr/include/c++/v1/optional:960:34: note: candidate function not viable: no known conversion from 'optional<...>' to 'const optional<...>' for object argument
>     constexpr value_type const&& value() const&&
>                                  ^
> (same error also occurs in isl_test_cpp17-checked.cc, which gets compiled next)
> ```
> 
> Naturally I tried to assign a C++17-compatible compiler at this formula. However, this update coincided with `gcc` upgrading to 12.1.0, which unfortunately depends on `isl` by itself… So I need to somehow make do with the `gcc 11.3.0_1` I currently have installed. I'm asking for help for modifying `Formula/isl.rb`, so that it compiles with minimal interference for all future upgrades. I want to add a build dependency block for High Sierra and below.
> 
> `Section break`
> 
> I now report what I have tried.
> 
> My version is migrated to `gcc@11` at version 11.3.0, so I `brew unlink gcc`, ~renamed `Cellar/gcc/` to `Cellar/gcc@11/` and then `Cellar/gcc@11/11.3.0_1/` to `Cellar/gcc@11/11.3.0/`, and finally `brew link gcc@11`~ made a symlink from gcc@11/11.3.0 to gcc/11.3.0_1 because moving the executables broke them.
> 
> * **Passing --cc=gcc-11 directly (viz. `brew install isl --cc=gcc-11`): Success.** But then I'll have to keep this old version of gcc around, or symlink the newest gcc every time, because `brew upgrade` keeps previous install options.
> * Passing --cc=gcc: Failure. Enigmatically, this doesn't require me to upgrade my `gcc` first. It still fails for compiler not supporting C++17's optionals (the same errors I pasted above), so it's probably not using Homebrew's gcc.
> * Adding `depends_on "gcc@11" => :build` to `isl.rb`: Failure in all circumstances below for the same error. Apparently the gcc it found was still the alias to system `clang`
>   
>   * Upgrading directly
>   * Setting `CC=gcc` in the interactive build environment
>   * Passing --cc=gcc
> * Adding `depends_on "gcc" => :build` instead: Failure. This time Homebrew wants to upgrade gcc first, which causes a circular dependency: `==> Installing dependencies for isl: gcc ==> Installing isl dependency: gcc ==> Installing dependencies for gcc: isl ==> Installing gcc dependency: isl ==> Installing dependencies for isl: gcc ==> Installing isl dependency: gcc (snip) ^C`
>   
>   * Pulling out the old `gcc.rb` file matching my installed version, so that brew doesn't hit this circular dependency bug: Failure with the same C++17 unsupported error.
> 
> All in all, the simplest way to solve this - adding a dependency block - doesn't seem to work, nor does the un-versioned --cc=gcc option. I don't like the working way left, so I'm asking for help.
> 
> `Section break`
> 
> For all the unfortunate souls stuck on High Sierra, who don't have any version of Homebrew's `gcc` yet, so can't even get the first gcc installed, here's a solution - pull out an old version of `isl`'s formula and install it first (I suppose you have dev tools already):
> 
> ```
> brew update         # Just to minimize the chance the formula gets updated in the mean time
> cd "$(brew --repository homebrew/core)"
> git checkout 912df90f851785d85995ea89db3ce125d7f4114e -- Formula/isl.rb         # This reverts to isl 0.24 which can be compiled with system clang
> brew install isl
> brew install gcc@11         # It's a long long build
> git checkout master -- Formula/isl.rb         # Get the newest formula back
> brew install isl --cc=gcc-11         # I hope some day we can pass --cc=gcc here, but not now
> brew install gcc        # now you get the latest gcc
> ```

If still not work, you need to set the environment variable `HOMEBREW_NO_INSTALL_FROM_API` to `1` before any `brew install` command.
